internal/lang/eval.go

```diff
@@ -275,14 +275,6 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 		return ctx, diags
 	}
 
-	// The reference set we are given has not been de-duped, and so there can
-	// be redundant requests in it for two reasons:
-	//  - The same item is referenced multiple times
-	//  - Both an item and that item's container are separately referenced.
-	// We will still visit every reference here and ask our data source for
-	// it, since that allows us to gather a full set of any errors and
-	// warnings, but once we've gathered all the data we'll then skip anything
-	// that's redundant in the process of populating our values map.
```

Comment:
I think this comment may no longer be relevant to the current implementation. 

Today,  we have the `.StaticValidReferences` call that validates all the static references, and the main loop here just retrieves things from the data source, and that retrieval is done solely based on the reference's subject, not the entire traversal.